### PR TITLE
Adds Mime exclusive traitor item: The Guide to Advanced Mimery

### PR DESCRIPTION
**Pull request by jessicamaybe**
## About the PR 
Adds new item to mime traitor uplink, The Guide to Advanced Mimery.

It is a spellbook that gives the mime 2 powers, Finger Guns and Invisible Blockade. It costs 20tc

Finger Guns has 3 uses, it shoots an invisible bullet that will do damage and render the victim temporarily muted.

Invisible blockade is like the normal mime invisible wall, except 3 tiles wide.

Also adds the option for spell books to let you keep your spells without needing to have the book in hand and also optionally lets spell books be single use.

**Screenshots**
image

**Changelog**
:cl:
- add: New Mime exclusive traitor item. The Guide to Advanced Mimery.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
Sorry, Uplink and uplinks related systems are currently under a freeze. As to not conflict with the ongoing refactor, either remove the catalog entry and add the ability some other way, or wait for the refactor to finish. Thank you.

**Comment by jessicamaybe** [comment]
> Sorry, Uplink and uplinks related systems are currently under a freeze. As to not conflict with the ongoing refactor, either remove the catalog entry and add the ability some other way, or wait for the refactor to finish. Thank you.

It can wait, there isn't any uplink related code just yaml so it's no big deal.

File: Resources/Locale/en-US/magic/spells-actions.ftl [hunk 1]
```diff
@@ -1,4 +1,6 @@
-﻿action-name-spell-rune-flash = Flash Rune
+﻿action-spellbook-crumble = The book crumbles
```
Comment (Veritius) [inline Resources/Locale/en-US/magic/spells-actions.ftl:1]: Could be something like `The book crumbles to dust`

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml [hunk 2]
```diff
@@ -111,3 +111,50 @@
       types:
         Piercing: 300
     ignoreResistances: true
+
+- type: entity
+  id: ProjectileFingerGun
+  name: mime bullet
+  description: insert raunchy finger bang joke
```
Comment (Veritius) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml:118]: Nobody'll see this but I still feel it should be properly capitalised, like:
`Insert raunchy finger bang joke here.`

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml [hunk 3]
```diff
@@ -111,3 +111,50 @@
...
+      soundHit:
+        path: /Audio/Weapons/Guns/Hits/bullet_hit.ogg
+    - type: StaminaDamageOnCollide
+      damage: 25
+    - type: TimedDespawn
+      lifetime: 10
+    - type: SolutionInjectOnCollide
+      transferAmount: 25
+    - type: InjectableSolution
+      solution: toxin
```
Comment (Veritius) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml:153]: why

File: Resources/Prototypes/Magic/forcewall_spells.yml [hunk 4]
```diff
@@ -13,3 +13,18 @@
   serverEvent: !type:InstantSpawnSpellEvent
     prototype: WallForce
     posData: !type:TargetInFront
+
+- type: instantAction
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Magic/forcewall_spells.yml:17]: does the mime not already have a power that's just making an invisible force wall?

Comment (Veritius) [reply Resources/Prototypes/Magic/forcewall_spells.yml:17]: they do but it's not 3 wide iirc

Comment (jessicamaybe) [reply Resources/Prototypes/Magic/forcewall_spells.yml:17]: Normal mime wall is one tile infront of them, the mimebook mime wall is 3 tiles wide instead

File: Content.Server/Magic/MagicSystem.cs [hunk 5]
```diff
@@ -144,6 +155,7 @@ private void OnInstantSpawn(InstantSpawnSpellEvent args)
                 var comp = EnsureComp<PreventCollideComponent>(ent);
                 comp.Uid = args.Performer;
             }
+
```
Comment (Elijahrane) [inline Content.Server/Magic/MagicSystem.cs:158]: whitespace

File: Content.Server/Magic/MagicSystem.cs [hunk 6]
```diff
@@ -115,7 +117,16 @@ private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHand
...
-        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+
+        if (!component.Persistent)
+            _actionsSystem.AddActions(ev.User, component.Spells, uid);
+        if (component.Persistent)
+            _actionsSystem.AddActions(ev.User, component.Spells, null);
+        if (component.SingleUse)
+        {
+            EntityManager.DeleteEntity(uid);
+            _popupSystem.PopupEntity(Loc.GetString("action-spellbook-crumble"), ev.User, Filter.Entities(ev.User), PopupType.Large);
```
Comment (Elijahrane) [inline Content.Server/Magic/MagicSystem.cs:128]: Should maybe be a caution popup?

File: Resources/Locale/en-US/magic/spells-actions.ftl [hunk 7]
```diff
@@ -1,4 +1,6 @@
-﻿action-name-spell-rune-flash = Flash Rune
+﻿action-spellbook-crumble = The book crumbles to dust.
```
Comment (Elijahrane) [inline Resources/Locale/en-US/magic/spells-actions.ftl:1]: {CAPITALIZE(THE($book))} crumbles to dust.

File: Content.Server/Magic/MagicSystem.cs (hunk 6 above)
Comment (Elijahrane) [reply Content.Server/Magic/MagicSystem.cs:128]: _popupSystem.PopupEntity(Loc.GetString("action-spellbook-crumble", ("book", uid)), ev.User, Filter.Entities(ev.User), PopupType.LargeCaution);

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml [hunk 8]
```diff
@@ -111,3 +111,50 @@
...
         Piercing: 300
     ignoreResistances: true
+
+- type: entity
+  id: ProjectileFingerGun
+  name: mime bullet
+  description: Insert raunchy finger bang joke here.
+  components:
+    - type: FlyBySound
+    - type: Clickable
```
Comment (Elijahrane) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml:121]: why does a projectile need to be clickable?

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml [hunk 9]
```diff
@@ -111,3 +111,50 @@
...
+    - type: SolutionInjectOnCollide
+      transferAmount: 25
+    - type: InjectableSolution
+      solution: toxin
+    - type: SolutionContainerManager
+      solutions:
+        toxin:
+          maxVol: 25
+          reagents:
+          - ReagentId: MuteToxin
```
Comment (Elijahrane) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml:159]: Silencer does the same thing and is more mime themed if anyone does some forensics

**Review by Elijahrane — CHANGES_REQUESTED** [review]

File: Content.Server/Magic/Components/SpellbookComponent.cs [hunk 10]
```diff
@@ -32,5 +32,19 @@ public sealed class SpellbookComponent : Component
...
+    [ViewVariables]
+    [DataField("isPersistent")]
+    public bool Persistent = false;
+
+    /// <summary>
+    /// If set to true the book will crumble after a single use
+    /// </summary>
+    [ViewVariables]
+    [DataField("singleUse")]
+    public bool SingleUse = false;
```
Comment (Veritius) [inline Content.Server/Magic/Components/SpellbookComponent.cs:47]: You can safely set these to `ViewVariables(VV.ReadWrite)`

File: Resources/Textures/Objects/Magic/magicactions.rsi/meta.json [hunk 11]
```diff
@@ -25,6 +25,12 @@
     {
       "name": "magicmissile"
     },
+    {
+      "name": "finger_guns"
+    },
+    {
+      "name": "invisible_blockade"
+    },
```
Comment (Veritius) [inline Resources/Textures/Objects/Magic/magicactions.rsi/meta.json:33]: Did you forget to add attribution for these?

Comment (jessicamaybe) [reply Resources/Textures/Objects/Magic/magicactions.rsi/meta.json:33]: They're from TG so it was already attributed

File: Resources/Prototypes/Catalog/uplink_catalog.yml [hunk 12]
```diff
@@ -426,6 +426,20 @@
       components:
       - SurplusBundle
 
+- type: listing
+  id: uplinkMimeBook
+  description: The classical two part series on how to further hone your mime skills.
+  productEntity: MimeSpellbook
+  cost:
+    Telecrystal: 20
```
Comment (EmoGarbage404) [inline Resources/Prototypes/Catalog/uplink_catalog.yml:434]: Is this even remotely worth 20 TC?  The gun is alright, but only 3 shots. The wall is good, but i don't think that it's good enough on its own to be worth this much.

File: Resources/Textures/Objects/Magic/magicactions.rsi/meta.json (hunk 11 above)
Comment (metalgearsloth) [reply Resources/Textures/Objects/Magic/magicactions.rsi/meta.json:33]: The copyright says cit.

Comment (jessicamaybe) [reply Resources/Textures/Objects/Magic/magicactions.rsi/meta.json:33]: Its got a link to citadel and tg

File: Content.Server/Magic/MagicSystem.cs [hunk 13]
```diff
@@ -115,7 +117,16 @@ private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHand
...
-        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+
+        if (!component.Persistent)
+            _actionsSystem.AddActions(ev.User, component.Spells, uid);
+        if (component.Persistent)
+            _actionsSystem.AddActions(ev.User, component.Spells, null);
+        if (component.SingleUse)
+        {
+            _popupSystem.PopupEntity(Loc.GetString("action-spellbook-crumble", ("book", uid)), ev.User, Filter.Entities(ev.User), PopupType.LargeCaution);
+            EntityManager.DeleteEntity(uid);
```
Comment (mirrorcult) [inline Content.Server/Magic/MagicSystem.cs:128]: make it actually crumble to dust, spawn ash on the ground

**Review by mirrorcult — COMMENTED** [review]
as emo mentioned 20tc is kinda steep. i'd be fine with 10-12 tc

File: Resources/Textures/Objects/Magic/magicactions.rsi/meta.json (hunk 11 above)
Comment (metalgearsloth) [reply Resources/Textures/Objects/Magic/magicactions.rsi/meta.json:33]: It should only be one source.
Either:
A) these states are from all of the sources in which case you should remove one anyway, or
B) these are mixed and need splitting out into a different rsi (or using RGA).

File: Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml [hunk 14]
```diff
@@ -111,3 +111,36 @@
       types:
         Piercing: 300
     ignoreResistances: true
+
+- type: entity
+  id: ProjectileFingerGun
+  name: mime bullet
+  description: Insert raunchy finger bang joke here.
```
Comment (metalgearsloth) [inline Resources/Prototypes/Entities/Objects/Weapons/Guns/Projectiles/magic.yml:118]: I think this is a bit much.

File: Resources/Prototypes/Magic/forcewall_spells.yml (hunk 4 above)
Comment (metalgearsloth) [reply Resources/Prototypes/Magic/forcewall_spells.yml:17]: I think it's better to have one or the other, not both.

File: Resources/Prototypes/Magic/projectile_spells.yml [hunk 15]
```diff
@@ -15,3 +15,20 @@
...
+  checkCanAccess: false
+  range: 60
+  sound: !type:SoundPathSpecifier
+    path: /Audio/Weapons/Guns/Gunshots/silenced.ogg
+  icon:
+    sprite: Objects/Magic/magicactions.rsi
+    state: finger_guns
+  serverEvent: !type:ProjectileSpellEvent
+    prototype: ProjectileFingerGun
+    posData: !type:TargetCasterPos
```
Comment (metalgearsloth) [inline Resources/Prototypes/Magic/projectile_spells.yml:34]: Should this just add guncomponent to the mime?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Pull request closed without merging**